examples/stress_tests/particles.rs

```diff
@@ -0,0 +1,226 @@
+        if keyboard_input.pressed(KeyCode::N) {
+            input |= INPUT_NOOP;
+        }
+
+        local_inputs.insert(*handle, input);
+    }
+
+    commands.insert_resource(LocalInputs::<Config>(local_inputs));
+}
+
+#[derive(Default, Reflect, Component, Clone, Copy, Deref, DerefMut)]
+struct Velocity(Vec3);
+
+impl std::hash::Hash for Velocity {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.0.x.to_bits().hash(state);
+        self.0.y.to_bits().hash(state);
+        self.0.z.to_bits().hash(state);
+    }
+}
+
+#[derive(Default, Reflect, Component, Clone, Copy, Deref, DerefMut)]
+struct Ttl(usize);
+
+type GameRng = rand_xoshiro::Xoshiro256PlusPlus;
+
+#[derive(Resource, Component, Clone, Deref, DerefMut)]
+struct ParticleRng(GameRng);
+
+fn main() -> Result<(), Box<dyn std::error::Error>> {
+    let args = Args::parse();
+    let num_players = args.players.len();
+    assert!(num_players > 0);
+
+    let desync_mode = match args.desync_detection_interval {
+        0 => DesyncDetection::Off,
+        interval => DesyncDetection::On { interval },
+    };
+
+    let mut session_builder = SessionBuilder::<Config>::new()
+        .with_num_players(num_players)
+        .with_desync_detection_mode(desync_mode)
+        .with_max_prediction_window(args.max_prediction)?
+        .with_input_delay(args.input_delay);
+
+    for (i, player_addr) in args.players.iter().enumerate() {
+        if player_addr == "localhost" {
+            session_builder = session_builder.add_player(PlayerType::Local, i)?;
+        } else {
+            let remote_addr: SocketAddr = player_addr.parse()?;
+            session_builder = session_builder.add_player(PlayerType::Remote(remote_addr), i)?;
+        }
+    }
+
+    for (i, spec_addr) in args.spectators.iter().enumerate() {
+        session_builder =
+            session_builder.add_player(PlayerType::Spectator(*spec_addr), num_players + i)?;
+    }
+
+    let socket = UdpNonBlockingSocket::bind_to_port(args.local_port)?;
+    let session = session_builder.start_p2p_session(socket)?;
+
+    App::new()
+        .add_plugins(GgrsPlugin::<Config>::default())
+        .set_rollback_schedule_fps(args.fps)
+        .add_systems(ReadInputs, read_local_inputs)
+        // SpriteBundle types
+        .rollback_component_with_clone::<Sprite>()
+        .rollback_component_with_clone::<Transform>()
+        .rollback_component_with_clone::<GlobalTransform>()
+        .rollback_component_with_clone::<Handle<Image>>()
+        .rollback_component_with_clone::<Visibility>()
+        .rollback_component_with_clone::<ComputedVisibility>()
+        // Also add our own types
+        .rollback_component_with_copy::<Velocity>()
+        .rollback_component_with_copy::<Ttl>()
+        .rollback_resource_with_clone::<ParticleRng>()
+        .checksum_component_with_hash::<Velocity>()
+        // todo: ideally we'd also be doing checksums for Transforms, but that's
```

Comment:
Agreed. Perhaps a `ComponentChecksumCustomHashPlugin` which would accept a function `fn<H: Hasher>(&C, &mut H)` could be used to retroactively add hashing to types which don't support it?